### PR TITLE
chore: pr for delete references

### LINKED ISSUE #37456: [Task] Switch delete branch method to super service
Notion Link: https://www.notion.so/Task-Switch-delete-branch-method-to-super-service-13cfe271b0e2806e8745ee942bb90df5

Notion Content:

### PR DESCRIPTION
**Pull request by sondermanish**
## Description
 - Implementation for delete references

Fixes #37456 

> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Git"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12374977893>
> Commit: ccb37fc00dba9b8eee62d150161343b1d18f4c89
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12374977893&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git`
> Spec:
> <hr>Tue, 17 Dec 2024 15:09:45 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to delete Git references, enhancing reference management capabilities.
	- Added error handling for branch deletion scenarios.
  
- **Changes**
	- Renamed existing method for clarity regarding Git reference creation.
	- Updated method signatures to reflect new functionality and improved naming conventions.
	- Added new methods to retrieve and set reference names in Git artifact metadata.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 1]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+    @Override
+    public Mono<? extends Artifact> deleteReference(
+            String baseArtifactId, String refName, ArtifactType artifactType, GitType gitType, RefType refType) {
+
+        if (refType == null) {
+            return Mono.error(new AppsmithException(AppsmithError.INVALID_PARAMETER, REF_TYPE));
+        }
+
+        if (!hasText(refName)) {
+            return Mono.error(new AppsmithException(AppsmithError.INVALID_PARAMETER, BRANCH_NAME));
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:584]: Nit, ref name

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 2]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
         return Mono.create(sink -> createBranchMono.subscribe(sink::success, sink::error, null, sink.currentContext()));
     }
 
+    @Override
+    public Mono<? extends Artifact> deleteReference(
+            String baseArtifactId, String refName, ArtifactType artifactType, GitType gitType, RefType refType) {
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:577]: Nit, can we use dto here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 3]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+        if (!hasText(refName)) {
+            return Mono.error(new AppsmithException(AppsmithError.INVALID_PARAMETER, BRANCH_NAME));
+        }
+
+        if (!hasText(baseArtifactId)) {
+            return Mono.error(new AppsmithException(AppsmithError.INVALID_PARAMETER, FieldName.ID));
+        }
+
+        GitArtifactHelper<?> gitArtifactHelper = gitArtifactHelperResolver.getArtifactHelper(artifactType);
+        AclPermission artifactEditPermission = gitArtifactHelper.getArtifactEditPermission();
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:592]: We ended up creating a lot of permission methods in the helper. Should we instead create a single getArtifactPermision method and access its methods directly for all our use cases?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 4]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+
+        GitArtifactMetadata baseGitMetadata = baseArtifact.getGitArtifactMetadata();
+        GitArtifactMetadata referenceArtifactMetadata = referenceArtifact.getGitArtifactMetadata();
+
+        GitHandlingService gitHandlingService = gitHandlingServiceResolver.getGitHandlingService(gitType);
+        GitArtifactHelper<?> gitArtifactHelper =
+                gitArtifactHelperResolver.getArtifactHelper(baseArtifact.getArtifactType());
+
+        // TODO: write a migration to shift everything to refName in gitMetadata
+        final String finalRefName = referenceArtifactMetadata.getBranchName();
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:618]: Can you in the mean time create an internal annotated getRefName method that returns branch name when ref name is null?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 5]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+                            .deleteGitReference(jsonTransformationDTO)
+                            .doFinally(signalType -> gitRedisUtils.releaseFileLock(baseArtifactId, TRUE))
+                            .flatMap(isReferenceDeleted -> {
+                                if (FALSE.equals(isReferenceDeleted)) {
+                                    return Mono.error(new AppsmithException(
+                                            AppsmithError.GIT_ACTION_FAILED,
+                                            " delete branch. Branch does not exists in the repo"));
+                                }
+
+                                if (referenceArtifact.getId().equals(baseArtifactId)) {
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:658]: Why don't we delete in this case, do you know?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 6]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+                                    return Mono.just(referenceArtifact);
+                                }
+
+                                return gitArtifactHelper
+                                        .deleteArtifactByResource(referenceArtifact)
+                                        .onErrorResume(throwable -> {
+                                            return gitAnalyticsUtils
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:665]: Should log error before transofrming

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java (hunk 5 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:658]: we don't delete baseArtifact, where would we keep the gitMetadata then?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 7]
```diff
@@ -572,6 +572,120 @@ public Mono<? extends Artifact> createReference(
...
+
+        if (finalRefName.equals(baseGitMetadata.getDefaultBranchName())) {
+            return Mono.error(new AppsmithException(
+                    AppsmithError.GIT_ACTION_FAILED, "delete ref", " Cannot delete default branch"));
+        }
+
+        Mono<? extends Artifact> deleteReferenceMono = gitPrivateRepoHelper
+                .isBranchProtected(baseGitMetadata, finalRefName)
+                .flatMap(isBranchProtected -> {
+                    if (!TRUE.equals(isBranchProtected)) {
```
[the remark this answers is not in the document]
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:629]: same stuff, even if we flip we are returning the error.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java (hunk 3 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:592]: do you mean a separate service?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java (hunk 2 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:577]: same gitRefDTO?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java (hunk 5 above)
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:661]: That's a bug then .. It's a stray object that we can never get rid of. Let's add a TODO to clean this up when we pull metadata out of artifact

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**